trinity/protocol/common/peer.py

```diff
@@ -0,0 +1,253 @@
+import operator
+import random
+from typing import (
+    Dict,
+    List,
+    Tuple,
+    Type,
+)
+
+from eth_typing import (
+    BlockNumber,
+    Hash32,
+)
+
+from eth_utils import (
+    decode_hex,
+    ValidationError,
+)
+from eth_utils.toolz import groupby
+
+from cancel_token import CancelToken
+
+from eth_keys import (
+    datatypes as eth_keys_datatypes,
+    keys,
+)
+
+from eth.chains.mainnet import MAINNET_NETWORK_ID
+from eth.chains.ropsten import ROPSTEN_NETWORK_ID
+from eth.constants import GENESIS_BLOCK_NUMBER
+from eth.rlp.headers import BlockHeader
+from eth.vm.base import BaseVM
+from eth.vm.forks import HomesteadVM
+
+from p2p.constants import (
+    DEFAULT_MAX_PEERS,
+)
+from p2p.kademlia import Address, Node
+from p2p.p2p_proto import DisconnectReason
+from p2p.peer import (
+    BasePeer,
+    BasePeerPool,
+    PeerConnection,
+)
+from p2p.exceptions import (
+    PeerConnectionLost,
+    NoConnectedPeers,
+)
+
+from trinity.db.header import BaseAsyncHeaderDB
+from trinity.exceptions import (
+    DAOForkCheckFailure,
+)
+
+from .constants import (
+    CHAIN_SPLIT_CHECK_TIMEOUT,
+)
+from .proto import ChainInfo
+
+
+class EVMPeer(BasePeer):
```

Comment:
I really dislike this name.  If anyone has any better ideas I'm all :ear: :ear: 

I'm also looking into whether we can do this with composition but I haven't quite seen how that would work.  Best stab looks roughly like:

- something for dealing with the boot sequence (dao check)
- container for metadata (`head_td`, `head_hash`).  Might be a service if it ends up responsible for tracking peer head.
- request/response API

I'm still wrapping my head around this...